File: decorator/decorator.py
#!/usr/bin/env python3

from abc import ABCMeta, abstractmethod


class Display(metaclass=ABCMeta):
    @abstractmethod
    def get_columns(self) -> int:
        pass

    @abstractmethod
    def get_rows(self) -> int:
        pass

    @abstractmethod
    def get_rowtext(self, row: int) -> str:
        pass

    def show(self) -> str:
        out = ""
        for i in range(self.get_rows()):
            out += self.get_rowtext(i)
        return out


class StringDisplay(Display):
    def __init__(self, string: str):
        self.string: str = string

    def get_columns(self) -> int:
        return len(self.string)

    def get_rows(self) -> int:
        return 1

    def get_rowtext(self, row: int) -> str:
        if row == 0:
            return self.string
        else:
            return ""


class Border(Display):
    def __init__(self, display: "Display"):
        self.display: "Display" = display


class SideBorder(Border):
    def __init__(self, display: "Display", char: str):
        super().__init__(display)
        self.border_char: str = char

    def get_columns(self) -> int:
        return 1 + self.display.get_columns() + 1

    def get_rows(self) -> int:
        return self.display.get_rows()

    def get_rowtext(self, row: int) -> str:
        return self.border_char + \
            self.display.get_rowtext(row) + self.border_char


class FullBorder(Border):

    def __init__(self, display: "Display"):
        super().__init__(display)

    def get_columns(self) -> int:
        return 1 + self.display.get_columns() + 1

    def get_rows(self) -> int:
        return 1 + self.display.get_rows() + 1

    def get_rowtext(self, row: int) -> str:
        if row == 0:
            return "+" + self.make_line("-", self.display.get_columns()) + "+"
        elif row == self.display.get_rows() + 1:
            return "+" + self.make_line("-", self.display.get_columns()) + "+"
        else:
            return "|" + self.display.get_rowtext(row - 1) + "|"

    def make_line(self, char: str, count: int) -> str:
        return char * count

File: decorator/decorator.go
package decorator

type display interface {
	getColumns() int
	getRows() int
	getRowText(row int) string
	Show(display display) string
}

type defaultDisplay struct {
	display
}

func (dd *defaultDisplay) Show(display display) string {
	str := ""
	for i := 0; i < display.getRows(); i++ {
		str += display.getRowText(i)
	}
	return str
}

type StringDisplay struct {
	*defaultDisplay
	str string
}

func NewStringDisplay(str string) *StringDisplay {
	return &StringDisplay{
		&defaultDisplay{},
		str,
	}
}

func (sd *StringDisplay) getColumns() int {
	return len(sd.str)
}

func (sd *StringDisplay) getRows() int {
	return 1
}

func (sd *StringDisplay) getRowText(row int) string {
	if row == 0 {
		return sd.str
	}
	return ""
}

type Border struct {
	*defaultDisplay
	display display
}

type SideBorder struct {
	*Border
	borderChar string
}

func NewSideBorder(display display, borderChar string) *SideBorder {
	return &SideBorder{
		&Border{display: display},
		borderChar,
	}
}

func (sb *SideBorder) getColumns() int {
	return 1 + sb.display.getColumns() + 1
}

func (sb *SideBorder) getRows() int {
	return sb.display.getRows()
}

func (sb *SideBorder) getRowText(row int) string {
	return sb.borderChar + sb.display.getRowText(row) + sb.borderChar
}
